internal/cli/snapshot_restore.go

```diff
@@ -0,0 +1,150 @@
+package cli
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/hashicorp/waypoint/internal/pkg/flag"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+	"github.com/posener/complete"
+	sshterm "golang.org/x/crypto/ssh/terminal"
+)
+
+type SnapshotRestoreCommand struct {
+	*baseCommand
+}
+
+// initWriter inspects args to figure out where the snapshot will be read from. It
+// supports args[0] being '-' to force reading from stdin.
+func (c *SnapshotRestoreCommand) initReader(args []string) (io.Reader, io.Closer, error) {
+	if len(args) >= 1 {
+		if args[0] == "-" {
+			return os.Stdin, nil, nil
+		}
+
+		f, err := os.Open(args[0])
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return f, f, nil
+	}
+
+	f := os.Stdin
+
+	if sshterm.IsTerminal(int(f.Fd())) {
+		return nil, nil, fmt.Errorf("stdin is a terminal, refusing to use (use '-' to force)")
+	}
+
+	return f, nil, nil
+}
+
+func (c *SnapshotRestoreCommand) Run(args []string) int {
+	// Initialize. If we fail, we just exit since Init handles the UI.
+	if err := c.Init(
+		WithArgs(args),
+		WithFlags(c.Flags()),
+		WithNoConfig(),
+	); err != nil {
+		return 1
+	}
+
+	client := c.project.Client()
+
+	r, closer, err := c.initReader(args)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to open output: %s", err)
+		return 1
+	}
+
+	if closer != nil {
+		defer closer.Close()
+	}
+
+	stream, err := client.RestoreSnapshot(c.Ctx)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to restore snapshot: %s", err)
+		return 1
+	}
+
+	err = stream.Send(&pb.RestoreSnapshotRequest{
+		Event: &pb.RestoreSnapshotRequest_Open_{
+			Open: &pb.RestoreSnapshotRequest_Open{},
+		},
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to send start message: %s", err)
+		return 1
+	}
+
+	// Write the data in smaller chunks so we don't overwhelm the grpc stream
+	// processing machinary.
+	var buf [1024]byte
+
+	for {
+		// use ReadFull here because if r is an OS pipe, each bare call to Read()
+		// can result in just one or two bytes per call, so we want to batch those
+		// up before sending them off for better performance.
+		n, err := io.ReadFull(r, buf[:])
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			err = nil
+		}
+
+		if n == 0 {
+			break
+		}
+
+		err = stream.Send(&pb.RestoreSnapshotRequest{
+			Event: &pb.RestoreSnapshotRequest_Chunk{
+				Chunk: buf[:n],
+			},
+		})
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to write snapshot data: %s", err)
+			return 1
+		}
+	}
+
+	_, err = stream.CloseAndRecv()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to receive snapshot start message: %s", err)
+		return 1
+	}
+
+	if r == os.Stdin {
+		c.ui.Output("Server data restored.")
+	} else {
+		c.ui.Output("Server data restored from '%s'.", args[0])
```

Comment:
I'd just add a bit more information here: by default this only _stages_ the restore data and the server still has to restart to complete it. So I'd let the user know of that reality, "restored" here isn't exactly right.

We may want to consider adding a flag to set the `Exit: true` field on the open message that crashes the server. Maybe `-exit` and document clearly that this will not RESTART IT on its own, we expect some other system to restart it.

Reply:
AH! Yes, I thought I was forgetting part of it! I'll do the notification and the exit flag now.